workflow/executor/pns/pns.go

```diff
@@ -0,0 +1,323 @@
+	return nil
+}
+
+// exitChroot exits chroot
+func (p *PNSExecutor) exitChroot() error {
+	if err := p.rootFS.Chdir(); err != nil {
+		return errors.InternalWrapError(err)
+	}
+	err := syscall.Chroot(".")
+	if err != nil {
+		return errors.InternalWrapError(err)
+	}
+	return nil
+}
+
+// CopyFile copies a source file in a container to a local path
+func (p *PNSExecutor) CopyFile(containerID string, sourcePath string, destPath string) (err error) {
+	destFile, err := os.Create(destPath)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		// exit chroot and close the file. preserve the original error
+		deferErr := p.exitChroot()
+		if err == nil && deferErr != nil {
+			err = errors.InternalWrapError(deferErr)
+		}
+		deferErr = destFile.Close()
+		if err == nil && deferErr != nil {
+			err = errors.InternalWrapError(deferErr)
+		}
+	}()
+	w := bufio.NewWriter(destFile)
+	err = p.enterChroot()
+	if err != nil {
+		return errors.InternalWrapError(err)
+	}
+
+	err = archive.TarGzToWriter(sourcePath, w)
+	if err != nil {
+		return errors.InternalWrapErrorf(err, "failed to targz to writer: %v", err)
+	}
+
+	return nil
+}
+
+func (p *PNSExecutor) WaitInit() error {
+	go p.pollRootProcesses(time.Minute)
+	// Secure a filehandle on our own root. This is because we will chroot back and forth from
+	// the main container's filesystem, to our own.
+	rootFS, err := os.Open("/")
+	if err != nil {
+		return errors.InternalWrapError(err)
+	}
+	p.rootFS = rootFS
+	return nil
+}
+
+// Wait for the container to complete
+func (p *PNSExecutor) Wait(containerID string) error {
+	mainPID, err := p.getContainerPID(containerID)
+	if err != nil {
+		return err
+	}
+	log.Infof("Main pid identified as %d", mainPID)
+	p.mainPID = mainPID
+	for pid, f := range p.pidFileHandles {
+		if pid == p.mainPID {
+			log.Info("Successfully secured file handle on main container root filesystem")
+			p.mainFS = f
+		} else {
+			log.Infof("Closing root filehandle for non-main pid %d", pid)
+			_ = f.Close()
+		}
+	}
+	if p.mainFS == nil {
+		log.Warn("Failed to secure file handle on main container's root filesystem. Output artifacts from base image layer will fail")
+	}
+
+	// wait for pid to complete
+	log.Infof("Waiting for main pid %d to complete", mainPID)
+	err = executil.WaitPID(mainPID)
+	if err != nil {
+		return err
+	}
+	log.Infof("Main pid %d completed", mainPID)
+	return nil
+}
+
+// pollRootProcesses will poll /proc for root pids (pids without parents) in a tight loop, for the
+// purpose of securing an open file handle against /proc/<pid>/root as soon as possible.
+// It opens file handles on all root pids because at this point, we do not yet know which pid is the
+// "main" container.
+// Polling is necessary because it is not possible to use something like fsnotify against procfs.
+func (p *PNSExecutor) pollRootProcesses(timeout time.Duration) {
+	log.Warnf("Polling root processes (%v)", timeout)
+	deadline := time.Now().Add(timeout)
+	for {
+		p.updateCtrIDMap()
```

Comment:
should executor panic if `updateCtrIDMap` returns error?

Reply:
this was changed to best effort/warn